File: atcoder_training/abc206/B_Savings.rs
//               B - Savings               
// ----------------------------------------
// 問題
// https://atcoder.jp/contests/abc206/tasks/abc206_b
// ----------------------------------------

// 二分探索

// use proconio::input;

fn main() {
    let mut n = String::new();
    std::io::stdin().read_line(&mut n).expect("Failed");
    let n: usize = n.trim().parse().unwrap();

    let (mut l, mut r) = (0, 1001001001);
    while (r - l) > 1 {
        let mid = (l + r) / 2;
        if mid * (mid+1) / 2 < n {
            l = mid;
        } else {
            r = mid;
        }
    }

    println!("{}", l+1);
}

File: contests/abc206/B_Savings.hs
--               B - Savings               
-- ----------------------------------------
-- 問題
-- https://atcoder.jp/contests/abc206/tasks/abc206_b
-- ----------------------------------------

{-
n <= k*(k+1)/2 < n+1 を満たすk
-}

main = readLn >>= (print >> solve)

solve n = n * n
-- solve ok ng n
--   | ng - ok < 1 = ok
--   | otherwise = 
--     if n <= mid * (mid+1) `div` 2
--     then solve mid ng n
--     else solve ok mid n
--   where mid = (ok + ng) `div` 2 